lib/util/cleverMerge.js

```diff
@@ -0,0 +1,88 @@
+/*
+	MIT License http://www.opensource.org/licenses/mit-license.php
+	Author Tobias Koppers @sokra
+*/
+
+"use strict";
+
+const mergeCache = new WeakMap();
+
+/**
+ * Merges two given objects and caches the result to avoid computation if same objects passed as arguments again.
+ * @example
+ * // performs cleverMerge(first, second), stores the result in WeakMap and returns result
+ * cachedCleverMerge({a: 1}, {a: 2})
+ * {a: 2}
+ *  // when same arguments passed, gets the result from WeakMap and returns it.
+ * cachedCleverMerge({a: 1}, {a: 2})
+ * {a: 2}
+ * @param {object} first first object
+ * @param {object} second second object
+ * @returns {object} merged object of first and second object
+ */
+const cachedCleverMerge = (first, second) => {
+	let innerCache = mergeCache.get(first);
+	if (innerCache === undefined) {
+		innerCache = new WeakMap();
+		mergeCache.set(first, innerCache);
+	}
+	const prevMerge = innerCache.get(second);
+	if (prevMerge !== undefined) return prevMerge;
+	const newMerge = cleverMerge(first, second);
+	innerCache.set(second, newMerge);
+	return newMerge;
+};
+
+/**
+ * Merges two objects. Objects are deeply clever merged.
+ * Arrays might reference the old value with "..."
+ * @param {object} first first object
+ * @param {object} second second object
+ * @returns {object} merged object of first and second object
+ */
+const cleverMerge = (first, second) => {
+	const newObject = Object.assign({}, first);
+	for (const key of Object.keys(second)) {
+		if (!(key in newObject)) {
+			newObject[key] = second[key];
+			continue;
+		}
+		const secondValue = second[key];
+		if (typeof secondValue !== "object" || secondValue === null) {
+			newObject[key] = secondValue;
```

Comment:
There should probably be `continue` here. If `secondValue` is not an object or null, then it will not be an Array and `firstValue` will equal `secondValue` in the else portion of the `if` below resulting in reassignment of `newObject[key]` to `secondValue`. The assignment of `newObject[key] = secondValue` currently occurs 4 times in this code. A cleaner approach might be to either introduce a variable called `newValue` which is initially set to `secondValue`, modified if needed, and then `newObject[key]` assigned to `newValue` at the end of the loop, OR negate the above `if` statement, wrap it around the one below, move this assignment to the end of the loop and add a `continue` after all other places where `newObject[key]` is set.